eternalcore-api/src/main/java/com/eternalcode/core/feature/jail/PrisonerService.java

```diff
@@ -0,0 +1,73 @@
+package com.eternalcode.core.feature.jail;
+
+import org.bukkit.command.CommandSender;
+import org.bukkit.entity.Player;
+
+import javax.annotation.Nullable;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+
+public interface PrisonerService {
+
+    /**
+     * Detains the player in jail.
+     *
+     * @param player     The player to detain.
+     * @param detainedBy The player who detained the player.
+     * @param duration   The duration of the detainment.
+     */
+    boolean detainPlayer(Player player, CommandSender detainedBy, @Nullable Duration duration);
+
+    /**
+     * Releases the player from jail.
+     *
+     * @param player     The player to release.
+     * @param releasedBy The player who released the player.
+     */
+    boolean releasePlayer(Player player, @Nullable CommandSender releasedBy);
+
+    /**
+     * Releases all players from jail.
+     */
+    void releaseAllPlayers();
+
+    /**
+     * Returns Set of jailedPlayers, which has string values of:
+     * Name of player
+     * Remaining time in jail
+     * Name of player who jailed them
+     */
+    Set<JailedPlayer> getJailedPlayers();
+
+    /**
+     * Returns true if command is allowed in jail.
+     * Ex. /help, /msg
+     *
+     * @param command The command to check.
+     */
+    boolean isCommandAllowed(String command);
+
+    /**
+     * Returns true if player is jailed.
+     *
+     * @param player The player to check.
+     */
+    boolean isPlayerJailed(UUID player);
+
+    /**
+     * Returns Map of prisoners, which has:
+     * A UUID of player
+     * Prisoner object
+     */
+    Collection<Prisoner> getPrisoners();
+
+    /**
+     * Returns true if player is prisoner.
+     *
+     * @param player The player to check.
+     */
+    boolean isNotInPrison(UUID player);
+}
```

Comment:
Here it's a bit mediocre in terms of the solution. When someone expects the API to cancel something, they usually don't want the API interface to control what happens after the cancellation (i.e., sending cancellation messages). It used to be fine before.